src/zero/api/handlers.clj

```diff
@@ -55,10 +55,17 @@
         (if-let [jwt (some-> request :oauth2/access-tokens :google :id-token
                              decode-jwt valid?)]
           (handler (assoc request :jwt jwt))
-          (-> (response/response {:message "Unauthorized"})
-              (response/header "WWW-Authenticate" "Bearer realm=API access")
-              (response/content-type "application/json")
-              (response/status 401)))))))
+          (if-let [bearer-token (some-> request
+                                        (response/get-header "authorization")
+                                        (str/split #" ")
+                                        last
+                                        decode-jwt
+                                        valid?)]
+            (handler (assoc request :jwt bearer-token))
+            (-> (response/response {:message "Unauthorized"})
+                (response/header "WWW-Authenticate" "Bearer realm=API access")
+                (response/content-type "application/json")
+                (response/status 401))))))))
```

Comment:
This would be better as two token sources validated the same way
instead of two completely distinct extraction and validation paths.

Reply:
^ Does that look better @tbl3rd?

Reply:
Better, but howzabout: (untested)
``` clojure
(letfn [(valid? [{:keys [payload] :as jwt}]
              (let [{:keys [aud exp hd iss]} payload]
                (and (= hd  "broadinstitute.org")
                     (= iss "https://accounts.google.com")
                     (= aud (get-in oauth2-profiles [:google :client-id]))
                     (> exp (quot (System/currentTimeMillis) 1000))
                     jwt)))
            (token [request]
              (get-in request [:oauth2/access-tokens :google :id-token]
                      (some-> request
                              (response/get-header "authorization")
                              (str/split #" ")
                              last)))]
      (if-let [jwt (some-> request token decode-jwt valid?)]
        (handler (assoc request :jwt jwt))
        (-> (response/response request)
            (response/header "WWW-Authenticate" "Bearer realm=API access")
            (response/content-type "application/json")
            (response/status 401))))
```
Maybe just return `request` for debugging.
Worried that no one will be looking for a `:message` anyway.
I also worry about the `split`, but that's probably OK.

Reply:
Indentation looks messed up, so please check the syntax in a real editor.

Reply:
Or you could `assoc-in` an `"Unauthorized"` result message.

Reply:
Thanks for the suggestion -- this looks much clearer now! :octocat: The only thing is I get a 500 error when I try to return the `request` as the response body

Reply:
Weird. Ah! Try something like this if your up for it.
``` clojure
{:request request :response {:message "Unauthorized"}}
```
Or … just merge it and maybe we'll take a later crack at it.

Reply:
Ring is probably trying trying to interpret the request as a response.

Reply:
Weirdly even including `{:request request}` causes a 500 error. I thought the `request` was just a map 🤔 Anyways I think we can revisit this later! Thanks for your help 💯